### ISSUE TITLE
Access to raw code objects

### ISSUE THREAD

**Issue by joonis** [dashingsoft/pyarmor#11]
I  obfuscated the following Python script.

**examples/test/mymod.py:**
```python
from __future__ import division, absolute_import, print_function, unicode_literals

def func2():
    print('func2')

def func3():
    print(err)
```

**Build pyarmor:**
```
python pyarmor.py init --src=examples/test --entry=mymod.py projects/test

python pyarmor.py config --disable-restrict-mode=1 projects/test

cd projects/test
./pyarmor build
```

**In Python shell:**
```
>>> import mymod, dis
>>> dis.dis(mymod)
Disassembly of func2:
  4     >>    0 LOAD_GLOBAL              0 (print)
              3 LOAD_CONST               1 (u'func2')
              6 CALL_FUNCTION            1
              9 POP_TOP             
             10 LOAD_CONST               0 (None)
             13 RETURN_VALUE        
             14 NOP                 
             15 NOP                 
             16 NOP                 
             17 NOP                 
             18 NOP                 
             19 NOP                 
             20 NOP                 
             21 NOP                 
             22 NOP                 
             23 NOP                 
             24 JUMP_ABSOLUTE            0

Disassembly of func3:
  7     >>    0 LOAD_GLOBAL              0 (print)
              3 LOAD_GLOBAL              1 (err)
              6 CALL_FUNCTION            1
              9 POP_TOP             
             10 LOAD_CONST               0 (None)
             13 RETURN_VALUE        
             14 NOP                 
             15 NOP                 
             16 NOP                 
             17 NOP                 
             18 NOP                 
             19 NOP                 
             20 NOP                 
             21 NOP                 
             22 NOP                 
             23 NOP                 
             24 JUMP_ABSOLUTE            0
```

Is it the correct behaviour that you have access to the raw code objects of a module?
So, what is the advantage of using pyarmor? Thanks.

**Comment by jondy** [comment]
If you restrict mode is enable, that is the default behavior, there is a limit:
* Obfuscated script can't imported from non-obfuscated scripts.
That is, you CAN NOT disassemble obfuscated module "mymod" by "dis", because no any other script can import "mymod" out of obfuscated scripts.

Refer to [Restricted mode](https://github.com/dashingsoft/pyarmor/blob/master/src/user-guide.md#restrict-mode)

But if you disable restrict mode, you can import obfuscated scripts from other script.  The byte code of code object will be restored after it's first called, so you can disassemble it.

**Comment by joonis** [comment]
Well, in this case the Odoo example is more or less useless.

https://github.com/dashingsoft/pyarmor/blob/master/src/user-guide.md#obfuscate-odoo-module

Because I can even use the raw bytecode (.pyc) rather than encrypting in unrestricted mode to get the same level of obfuscation.

Pyarmor should wrap all functions and methods:
```
def wrap(func):
    def wrapper(*args, **kwargs):
        return func(*args, **kwargs)
    wrapper.__module__ = func.__module__
    wrapper.__name__ = func.__name__
    wrapper.__doc__ = func.__doc__
    wrapper.__dict__.update(func.__dict__)
    return wrapper
```
Any downsides to this?

**Comment by jondy** [comment]
The only downside of wrap all functions is to reduce the performance, but it's a way. I'll try to implement it and test the performance. If it's acceptable, add this feature in next minor release. Thanks.

**Comment by jondy** [comment]
The workaround may be like this
1. Add a built function `__wraparmor__`
   It will restore func_code before call function and obfuscate func_code after it returns.

2. Add a decorator `wraparmor`
```
def wraparmor(func):
    def wrapper(*args, **kwargs):
        __wraparmor__(func.func_code)
        result = func(*args, **kwargs)
        __wraparmor__(func.func_code)
        return result
    wrapper.__module__ = func.__module__
    wrapper.__name__ = func.__name__
    wrapper.__doc__ = func.__doc__
    wrapper.__dict__.update(func.__dict__)
    return wrapper
```

3. Compile python scripts to ast node, and add decorator "wraparmor" for each function and class method.

4. Obfuscate wrapped python scripts.

**Comment by joonis** [comment]
But it should go into a try/finally:

```python
def wraparmor(func):
    def wrapper(*args, **kwargs):
        __wraparmor__(func.func_code)
        try:
            return func(*args, **kwargs)
        finally:
            __wraparmor__(func.func_code)
    wrapper.__module__ = func.__module__
    wrapper.__name__ = func.__name__
    wrapper.__doc__ = func.__doc__
    wrapper.__dict__.update(func.__dict__)
    return wrapper
```
With the downside, that we loose the function signatures...

**Comment by jondy** [comment]
Nice. And I read the source of built-in decorator `classmethod `and `staticmethod` in file `Objects/funcobject.c`. Maybe it's another way to write a quick decorator.

**Comment by jondy** [comment]
The final decorator will be
```python
def wraparmor(func):
    def wrapper(*args, **kwargs):
         __wraparmor__(func)
         try:
             return func(*args, **kwargs)
         finally:
             __wraparmor__(func, 1)
    wrapper.__module__ = func.__module__
    wrapper.__name__ = func.__name__
    wrapper.__doc__ = func.__doc__
    wrapper.__dict__.update(func.__dict__)
    func.__refcalls__ = 0
    return wrapper
```
Add an attribute `__refcalls__` to fix recursive call and multi-threads problem.

**Comment by jondy** [comment]
Released in [v3.7.0](https://github.com/dashingsoft/pyarmor/releases/tag/v3.7.0)

Here is the basic usage: [Use decorator to protect code objects when disable restrict mode](https://github.com/dashingsoft/pyarmor/blob/master/src/user-guide.md#use-decorator-to-protect-code-objects-when-disable-restrict-mode)

Here is an example: [Protect module with decorator "wraparmor"](https://github.com/dashingsoft/pyarmor/blob/master/src/user-guide.md#protect-module-with-decorator-wraparmor)

**Comment by joonis** [comment]
Seems to work. Even the debugger crashes ;) but this needs further testing...

One thing to note: If an exception occurs in `func` someone could go back in the stack frames and access the local variables of `func`. To prevent this you could modify the wrapper function:

```python
def wraparmor(func):
    def wrapper(*args, **kwargs):
        __wraparmor__(func)
        try:
            return func(*args, **kwargs)
+       except Exception as err:
+           raise err
        finally:
            __wraparmor__(func, 1)
    wrapper.__module__ = func.__module__
    wrapper.__name__ = func.__name__
    wrapper.__doc__ = func.__doc__
    wrapper.__dict__.update(func.__dict__)
    func.__refcalls__ = 0
    return wrapper
```
This way it is possible to pass the corresponding frame. It has the downside of a wrong location of the exception and makes debugging more difficult. But if you have some secret variables in a particular function someone could use this decorator instead.

**Comment by joonis** [comment]
Users must also know that globals are mutable but locals are not. So you should always wrap the whole module within a function and just make public functions global:

```python
@wraparmor
def __load__():

    @wraparmor
    def my_private_func():
        pass

    global my_public_func
    @wraparmor
    def my_public_func():
        pass

__load__()
```

**Comment by joonis** [comment]
One more thing: Would it be possible to allow the import of the entry module only, but restrict the direct import of any other module?

**Comment by jondy** [comment]
Good idea! The workaround would be **Only entry module can run "pyarmor_runtime()"**

**Rationale:**
* pyarmor_runtime will call c function "init_runtime"
* init_runtime will check the current code object in the stack frame whether it's entry module or not
* The identify information of entry module will be stored in the project license (license.lic)

If there is more than one entry, end users just generate different project license for each entry script.

Why didn't I think of it, it's terrific! After wall, the way to use decorator is somewhat complicated, and just as you have mentioned, need more and more tests.

**Comment by joonis** [comment]
Next issue: Only encrypted modules must be allowed to call `__wraparmor__`. Otherwise everyone could just decrypt an encrypted function.

**Comment by jondy** [comment]
From v3.7.1, `__wraparmor__` only can be called from decorator `wraparmor`, it will check like this
```c
static PyObject *wrapcaller = NULL;
static PyObject*
__wraparmor__(PyObject *self, PyObject *args)
{
    PyObject *co_code = PyEval_GetFrame()->f_code->co_code;
    // First time call, set wrapcaller to code object of function wrapper in decorator
    if (!wrapcaller) 
        wrapcaller = co_code;
    // If it's not called from the same code object, return NULL, and no exception set
    else if (wrapcaller != co_code)
       return NULL;
    // Go on ...
```

**Comment by jondy** [comment]
Adding checkpoints in c function `init_runtime` can **NOT** avoid importing obfuscated module from other scripts, it only works when entry script is module `__main__`. So it's not apply to the case of odoo module. I think it over, and I found it's difficult to deal with this case.  For now, use decorator "wraparmor" is one alternative solution.

**Comment by joonis** [comment]
Great, `wraparmor` seems to work reliable now.

**Comment by joonis** [comment]
Is it somehow possible to prevent a wrapped function dectrypting itself by just calling it?
Because a wrapped function should be decryptable only by `__wraparmor__`...

**Comment by jondy** [comment]
But I think if wrapped function will be encrypted again as soon as it returns in any way, it should not be a problem.

**Comment by joonis** [comment]
Well, I think you are right. Just triggered an exception, got the respective wrapper frame, the wrapped function in its locals and tried to call it directly. It did not work. So everthing seems to be fine. But shouldn't it be callable this way?

**Comment by jondy** [comment]
I have refined `__wraparmor__` in v3.8.1, it will crash to access original func_code out of decorator.

**Comment by joonis** [comment]
Unfortunately ipython crahes as well now, since it tries to access the func_code apparently.
Anyway, no problems with v3.8.0. Maybe Pyarmor should just raise a SystemError again, instead causing a segfault. What do you think?

**Comment by jondy** [comment]
It's not possible to raise SystemError because `__wraparmor__` just modify some members of func_code when function returns.  One way is to add an option to disable this constraint.

**Comment by joonis** [comment]
Patching IPython works as well: https://stackoverflow.com/a/28758396
It's the better solution I think.

**Comment by joonis** [comment]
The problem is accessing `co_varnames` of the wrapped code object.
Maybe there is a possibility to leave this attribute untouched?

**Comment by jondy** [comment]
It's no problem.

**Comment by joonis** [comment]
What exactly is no problem? ;)

**Comment by jondy** [comment]
Leave `co_varname` untouched, :)

**Comment by joonis** [comment]
Ok. Because at time it causes a segfault...

**Comment by joonis** [comment]
Jondy, leave it as is... There are still other issues. I think it is easier to patch IPython or the inspect module instead. Thanks anyway!

**Comment by joonis** [comment]
Well, with Django we have the same problems and it isn't that easy to fix.
So after all we need access to `co_varnames`.

**Comment by joonis** [comment]
I don't know the internals, but could we reduce the values to the number of parameters of the function?

```
def func(a, b):
    c = a + b
    return c
```
Here `co_varnames` usually is `('a', 'b', 'c')` (parameter + local names).
Maybe we could just return `('a', 'b')` if it is still executable then...

**Comment by joonis** [comment]
Same with `frame.f_locals`. Here we should return an empty dict.

**Comment by jondy** [comment]
Python interpreter maybe doesn't work if change content of `co_varnames`. 

Maybe the best way is to set `frame.f_locals` to an empty dict when raise exception, so original `func` can not be got from frame.

**Comment by joonis** [comment]
Sounds reasonable.

**Comment by jondy** [comment]
In v3.8.3, remove freevar `func` from `frame.f_locals` when raise exception in decorator `wraparmor`, so original function can not be got from exception traceback. Besides, only two members `co_consts` and `co_code`  are be changed, all the other members don't touched.

**Comment by joonis** [comment]
Now I get: SystemError: ../Objects/cellobject.c:36: bad argument to internal function
Don't know if you have an idea. Otherwise I try to track it down.

**Comment by joonis** [comment]
The previous error I got in Py2. With Py3 I get:
NameError: free variable 'func' referenced before assignment in enclosing scope

**Comment by jondy** [comment]
Bad news. In order to clear freevar `func`, I modified `frame.f->localsplus`, and call `PyCell_Set(var, NULL)`. The error in your Py2 should be made by `PyCell_Set`, but it works in my machine.

The error in Py3 I need to check code.

**Comment by jondy** [comment]
I guess the reason is wrong offset of freevar `func` in frame. In this version, pyarmor assume the last freevar is 'func', and the previous freevar is `err` in decorator `wraparmor`.  So does the decorator `wraparmor` in your script like this

```Python
def wraparmor(func):
    def wrapper(*args, **kwargs):
        __wraparmor__(func)
        try:
            return func(*args, **kwargs)
        except Exception as err:
            raise err
        finally:
            __wraparmor__(func, 1)
    wrapper.__module__ = func.__module__
    wrapper.__name__ = func.__name__
    wrapper.__doc__ = func.__doc__
    wrapper.__dict__.update(func.__dict__)
    func.__refcalls__ = 0
    return wrapper
```

And could you show the disassembling code object of `wrapper`,  assume the script name is `test.py`, the following commands will disassemble function `wrapper`
```bash
python -c'import dis, test
dis.dis(test.wraparmor.__code__.co_consts[2])'

python3 -c'import dis, test
dis.dis(test.wraparmor.__code__.co_consts[2])'
```

**Comment by jondy** [comment]
The code to clear `func` is
```c
        PyObject *frame = PyEval_GetFrame();
        PyObject **p = frame->f_valuestack;

        if (*(p - 2)) {
          p --;
          // Clear 'func`
          if (PyCell_Set(*p, NULL) == -1) {
          }
          p --;
          // Clear `err`
          if (PyCell_Set(*p, NULL) == -1) {
          }
```

**Comment by joonis** [comment]
Py2 dis:
```
Traceback (most recent call last):
  File "mytest.py", line 18, in <module>
    dis.dis(wraparmor.__code__.co_consts[2])
  File "/usr/lib/python2.7/dis.py", line 49, in dis
    type(x).__name__
TypeError: don't know how to disassemble int objects
```
Py3 dis:
```
Traceback (most recent call last):
  File "/usr/lib/python3.4/dis.py", line 26, in _try_compile
    c = compile(source, name, 'eval')
  File "<dis>", line 1
    wraparmor.<locals>.wrapper
              ^
SyntaxError: invalid syntax

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "mytest.py", line 18, in <module>
    dis.dis(wraparmor.__code__.co_consts[2])
  File "/usr/lib/python3.4/dis.py", line 59, in dis
    _disassemble_str(x, file=file)
  File "/usr/lib/python3.4/dis.py", line 348, in _disassemble_str
    disassemble(_try_compile(source, '<dis>'), file=file)
  File "/usr/lib/python3.4/dis.py", line 28, in _try_compile
    c = compile(source, name, 'exec')
  File "<dis>", line 1
    wraparmor.<locals>.wrapper
              ^
SyntaxError: invalid syntax
```

**Comment by jondy** [comment]
Try to disassemble `wraparmor.__code__.co_consts[1]` or `wraparmor.__code__.co_consts[3]`, the code object of `wrapper' is in co_consts somewhere.

**Comment by jondy** [comment]
@joonis I disassemble your decorator, remove freevar `shift_error` from `_call`, v3.8.3 should work for your custom decorator.

**Comment by joonis** [comment]
It does only work with the base version of wraparmor:

```python
        def wraparmor(func):
            def wrapper(*args, **kwargs):
                __wraparmor__(func)
                try:
                    return func(*args, **kwargs)
                except Exception as err:
                    raise err
                finally:
                    __wraparmor__(func, 1)
            wrapper.__module__ = func.__module__
            wrapper.__name__ = func.__name__
            wrapper.__doc__ = func.__doc__
            wrapper.__dict__.update(func.__dict__)
            func.__refcalls__ = 0
            return wrapper
```

**Comment by jondy** [comment]
Maybe it's better to leave `frame.f_locals` as it is, do not touch it by this way. So the decorator can control everything. Even if end user can access `frame.f_locals`, there are only the variables defined in decorator, and co_consts of `func` can't be accessed, the bytecode of `func` is obfuscated.

**Comment by joonis** [comment]
But it's not the case if you are doing it this way:

```python
        def wraparmor(func):
            def wrapper(*args, **kwargs):
                __wraparmor__(func)
                try:
                    return func(*args, **kwargs)
                except Exception as err:
                    raise # <--------------------------------
                finally:
                    __wraparmor__(func, 1)
            wrapper.__module__ = func.__module__
            wrapper.__name__ = func.__name__
            wrapper.__doc__ = func.__doc__
            wrapper.__dict__.update(func.__dict__)
            func.__refcalls__ = 0
            return wrapper
```
Here you have additional frames of at least one encrypted function.
But I think you are right. If you have some secret variables in a function, you have to shift the exception via `raise err`, otherwise `raise` is sufficient to keep it debugable. Finally this was the intension by shifting the exception ;)

So you are going to undo changes introduced in v3.8.1, right?

**Comment by jondy** [comment]
Undo change to v3.8.2, only a little change: only obfuscate `co_consts` and `co_code` of `func`, do not touch all other members.

**Comment by joonis** [comment]
Well, clearing `f_locals` should be quite easy.
`__wraparmor__(func, 1)` can do the following (regardless of an exception):
```
func_frame.f_locals.clear()
PyFrame_LocalsToFast(func_frame, int 1)
```

**Comment by jondy** [comment]
Nice code. But if clear `f_locals` when no exception, it will complaint
```
NameError: free variable 'func' referenced before assignment in enclosing scope
```

**Comment by joonis** [comment]
You are right. In addition changes made with `PyFrame_LocalsToFast` are not persistent.

**Comment by joonis** [comment]
Finally I found a solution:
```python
        def wraparmor(func):
            def wrapper(*args, **kwargs):
                __wraparmor__(func)
                try:
                    return func(*args, **kwargs)
                except Exception as err:
                    ##############################
                    # This should be implemented by Pyarmor
                    tb = err.__traceback__.tb_next
                    while tb:
                        tb.tb_frame.clear()
                        tb = tb.tb_next
                    ##############################
                    raise
                finally:
                    __wraparmor__(func, 1)
            wrapper.__module__ = func.__module__
            wrapper.__name__ = func.__name__
            wrapper.__doc__ = func.__doc__
            wrapper.__dict__.update(func.__dict__)
            func.__refcalls__ = 0
            return wrapper
```
Unfortunately the functionality of `clear()` is only public available in Py3 but implemented in Py2 as well:
- https://github.com/python/cpython/blob/3.6/Objects/frameobject.c#L525
- https://github.com/python/cpython/blob/2.7/Objects/frameobject.c#L557

Therefore this should be done by Pyarmor altogether.
The following method definition is missing in Py2 and must be applied by Pyarmor:
https://github.com/python/cpython/blob/3.6/Objects/frameobject.c#L558

**Comment by jondy** [comment]
As your idea, but a simplified implementation, I clears `f_locals` from traceback in `__wraparmor__(func, 1)` as the following way
```c
        PyObject *ts = PyThreadState_Get();
        PyObject *exc_value = *(PyObject**)((char*)ts + ts_exc_value_offset);
        if (exc_value) {
          PyObject *tb = *(PyObject**)((char*)ts + ts_exc_traceback_offset);
          while (tb) {
            PyObject *tb_frame = *(PyObject**)((char*)tb + tb_frame_offset);
            clear_frame_locals(tb_frame);
            tb = *(PyObject**)((char*)tb + tb_next_offset);
          }
        }
```
But I found the frames inner `wrapper` are gone, does it work as you expected? I build a temporary version for this feature

http://pyarmor.dashingsoft.com/downloads/platforms/linux_x86_64/temp/_pytransform.so
md5sum: be34841b2c79c39bb61481411567f531

**Comment by joonis** [comment]
Yes `frame.clear()` was introduced with Py34. Before this version it must be implemented as for Py2.

And yes, in Py3 exceptions are only available within the except block. So your build doesn't work. What about the following:

```python
                tb = None
                try:
                    return func(*args, **kwargs)
                except Exception as err:
                    tb = sys.exc_info()[2].tb_next
                    raise
                finally:
                    __wraparmor__(func, tb, 1)
```

**Comment by joonis** [comment]
Well, just calling `tp_clear` from `PyFrameType` should be sufficient in both, Py2 and 3.
In Py2 we have to ensure manually that the traceback frame isn't active anymore.
But this should be the case for `sys.exc_info()[2].tb_next` and all inner traceback objects.

**Comment by jondy** [comment]
I build a temporary version with `tp_clear`, it need an extra argument `tb` from decorator as your advice. The downside is that it will crash if try to access `frame.f_locals` in except block before python34. From Py34, it seems work well.

http://pyarmor.dashingsoft.com/downloads/platforms/linux_x86_64/temp/_pytransform.so
md5sum: 43e5aec6ac8421e74d12d010505fe6b0

**Comment by joonis** [comment]
`__wraparmor__(func, tb, 1)`
TypeError: function takes at most 2 arguments (3 given)

**Comment by joonis** [comment]
Ok, maybe I got a cached version. Py3 seems to work. Py2 segfaults :(

**Comment by joonis** [comment]
As you mentioned, Py2 segfaults if accessing `f_locals`, otherwise it works as well.

**Comment by joonis** [comment]
I think the problem is that `PyFrame_FastToLocals` has been replaced with `PyFrame_FastToLocalsWithError` in Py34+, since the segfault occurs in `PyFrame_FastToLocals` in Py2.

So, isn't it possible to change the getter function for `f_locals` after clearing to something like the following:
```
static PyObject *
frame_getlocals_post_clear(PyFrameObject *f, void *closure)
{
    /* PyFrame_FastToLocals(f);  <-- This causes the segfault */
    Py_INCREF(f->f_locals);
    return f->f_locals;
}
```
And then change the getter for all inner frames to something like:
```
    {"f_locals",        (getter)frame_getlocals_post_clear, NULL, NULL},
```

**Comment by joonis** [comment]
Or is it just the access of `f->f_locals`?
Maybe it is just enough to reset it afterwards for each frame: `f->f_locals = PyDict_New();`

**Comment by joonis** [comment]
I think this is the one and only solution: In Python 3.4+ we will `clear()` all frames.
In previous versions we set `f->f_locals` of all frames to an instance of the following dict subclass:
```python
class LocalsDict(dict):
    
    def __init__(self, *args, **kwargs):
        dict.__init__(self)

    def __setitem__(self, key, value):
        pass
    
    def __delitem__(self, key):
        pass
```
This way `f_locals` is always empty.

**Comment by jondy** [comment]
I change the setter of `f_locals` as the following way, it's easier
```c
static PyObject *
frame_getlocals(PyObject *f, void *closure)
{
  PyObject *d = Py_BuildValue("{}");
  Py_IncRef(d);
  return d;
}

static inquiry tp_clear = NULL;

    // Change setter to my own only once
    if (!tp_clear) {
      PyObject *frame_getsetlist = *((PyObject**)((char*)(frame->ob_type) + tp_getset_offset));
      *(PyObject**)((char*)frame_getsetlist + sizeof(char*)) = (PyObject*)frame_getlocals;
      tp_clear = *((inquiry*)((char*)(frame->ob_type) + tp_clear_offset));
    }
```

It works well, all the f_locals are empty. And I have a question: if `f_locals` always is empty, does `tp_clear` is required?

Maybe the best solution is to change the setter of `f_locals` to an empty dictionary in all python versions, do not touch `tp_clear` any more, :)

**Comment by jondy** [comment]
The downside is that local values are still in frame stack, although it can't be got from `frame.f_locals`, it could be got by some other way. So `tp_clear` is still required.

**Comment by joonis** [comment]
Yes, overall `tp_clear` is the better way. I mean, there is always a way to get the data, but this makes it really difficult.

Another point: Isn't `PyGetSetDef` set type-wide? Then all frames would be affected by this change. In this case it would be better to use the LocalsDict, isn't it?

**Comment by joonis** [comment]
Ok, just tried v3.8.5. Changing the getter globally affects all frames.
I don't know if want to do this. What do you think?

**Comment by joonis** [comment]
The cleanest solution would be the following:

Pyarmor defines the `LocalsDict` class and sets `f->f_locals` to an instance of this dict as soon as the frame of the encrypted function is created (thus at the beginning of the function call, regardless of an exception). Would this be possible? The advantage is, that even callbacks from encrypted code to outside code are secured of getting its `f_locals`. That's all.

Calling `frame.clear()` on exceptions can be done within the wrapper function, independently from Pyarmor.

**Comment by joonis** [comment]
Argh. What about something like this:
```c
static PyObject *
frame_getlocals(PyObject *f, void *closure)
{
    if (f->is_pyarmor) {
        PyObject *d = Py_BuildValue("{}");
        Py_IncRef(d);
        return d;
    }
    PyFrame_FastToLocals(f);
    Py_INCREF(f->f_locals);
    return f->f_locals;
}
```

**Comment by jondy** [comment]
One possible solution, let the global getter decide what to do
```c
static PyObject *
frame_getlocals(PyFrameObject *frame, void *closure)
{
  PyCodeObject *co = frame->f_code;
  int flag = co->co_flags;

  // CO_WRAP_FLAG will be set if byte code is obfuscated
  if (flag & CO_WRAP_FLAG) {
    PyObject *d = Py_BuildValue("{}");
    Py_IncRef(d);
    return d;
  }

  else
    return original_frame_getlocals(frame, closure);
}
```

**Comment by jondy** [comment]
Look, this is my code, :) 
It's same as your idea.
And when you type your idea, I'm typing my code.

**Comment by joonis** [comment]
Yes, it is :)
Since it is type-wide, it should be applied once, regardless of an exception. Great!

**Comment by joonis** [comment]
However, `frame.clear()` could be done in wrapper function as well. Whatever you prefer...

**Comment by jondy** [comment]
But `frame.clear()` is new in Python34. How about earlier python version?

**Comment by joonis** [comment]
I still thought we cannot clear in Py2, but since we have the getter it should be possible.
So you are right, Pyarmor should clear all frames on exceptions...

**Comment by joonis** [comment]
Why should it crash?

**Comment by jondy** [comment]
I means in c code it need to very careful. :)

I have a trip a few hours later, and will be back in 3 days. I'll implement this feature when I'm back. 

Just now, I find another issue,  I'm afraid I have no time to think of it now. For example
```
def foo():
  raise Exception('something is wrong')

@wraparmor
def protected_method():
  foo()
```
When exception happened,  `frame` of foo will be clear. But CO_WRAP_FLAG is not set in its code object. So `frame.f_locals' will be crashed because it's the original getter takes effect.

**Comment by joonis** [comment]
Why should the frame of foo be cleared? It's not encrypted. I will play around with this meanwhile...

My tests didn't crash for a fairly complex app. Only generator functions are not supported at time.
You could mark it as experimental and it can be disabled by passing `None` as traceback argument.

**Comment by joonis** [comment]
I see, of course all inner frames are cleared at time.
Maybe it makes sense to check for `CO_WRAP_FLAG` before clearing...

**Comment by jondy** [comment]
I think so too. Only clear those marked frame.

**Comment by joonis** [comment]
Alternatively you could set an extra `co_flag` for each code object when it is initially obfuscated.
Then you can check for this flag in the getter function...

**Comment by jondy** [comment]
In v3.8.6, `frame.f_locals` of `wrapper` and wrapped functions return an empty dictionary, and only their frames are clear. All the other frames changed nothing.

The decorator would be like this
```python
def wraparmor(func):
    func.__refcalls__ = 0
    def wrapper(*args, **kwargs):
         __wraparmor__(func)
         tb = None
         try:
             return func(*args, **kwargs)
         except Exception:
             tb = sys.exc_info()[2]
             raise
         finally:
             __wraparmor__(func, tb, 1)
    wrapper.__module__ = func.__module__
    ...
    return wrapper
```

When call `__wraparmor__(func, tb, 1)`
```c
static inquiry tp_clear = NULL;
static void* original_frame_getloals = NULL;

static PyObject*
__wraparmor__(PyObject *self, PyObject *args)
{
    ...
    // If tb is None, no frame will be clear, 
    // and the getter of frame.f_locals is not changed
    if (tb != None) {
        // Get frame of `wrapper`
        PyFrameObject *frame = PyEval_GetFrame();
        if (!tp_clear) {
            tp_clear = (frame->ob_type)->tp_clear;
            original_frame_getloals = *(void**)((char*)frame_getsetlist + sizeof(char*));
            *(void**)((char*)frame_getsetlist + sizeof(char*)) = (void*)frame_getlocals;
        }
        // Only clear wrapper and wrapped function
          while (tb) {
            PyObject *tb_frame = tb -> tb_frame;
            tp_clear(tb_frame);
            if (tb_frame != frame)
              break;
            tb = tb->tb_next;
          }
    }
}
```

The customized getter of `frame.f_locals`

```c
static PyObject *
frame_getlocals(PyObject *frame, void *closure)
{
  PyObject *f_code = *(PyObject**)((char*)frame + f_code_offset);
  PyObject *co_code = *(PyObject**)((char*)f_code + co_code_offset);
  int co_flags = *(int*)((char*)f_code + co_flags_offset);

  if ((co_flags & CO_WRAP_FLAG) || (co_code == wrapcaller)) {
    PyObject *d = Py_BuildValue("{}");
    Py_IncRef(d);
    return d;
  }

  return original_frame_getloals(frame, closure);
}
```

**Comment by joonis** [comment]
Well, I see.
```c
         // Only clear wrapper and wrapped function
          while (tb) {
            PyObject *tb_frame = tb -> tb_frame;
            tp_clear(tb_frame);
            if (tb_frame != frame)
              break;
            tb = tb->tb_next;
          }
```
But this clears exactly these two frames. What if I have a deeper traceback?
Can't you set just another `co_flag` on obfuscation for each code object (eg. `CO_PYARMOR_FLAG`).

Now initially set the getter:
```c
static PyObject *
frame_getlocals(PyObject *frame, void *closure)
{
  PyObject *f_code = *(PyObject**)((char*)frame + f_code_offset);
  PyObject *co_code = *(PyObject**)((char*)f_code + co_code_offset);
  int co_flags = *(int*)((char*)f_code + co_flags_offset);

  if (co_flags & CO_PYARMOR_FLAG) {
    PyObject *d = Py_BuildValue("{}");
    Py_IncRef(d);
    return d;
  }
  return original_frame_getloals(frame, closure);
}
static void* original_frame_getloals = *(void**)((char*)frame_getsetlist + sizeof(char*));
*(void**)((char*)frame_getsetlist + sizeof(char*)) = (void*)frame_getlocals;
```
And clearing the tracebacks in `__wraparmor__`:
```c
static inquiry tp_clear = NULL;
static void* original_frame_getloals = NULL;

static PyObject*
__wraparmor__(PyObject *self, PyObject *args)
{
    ...
    // If tb is None, no frame will be clear
    if (tb != None) {
        // Get frame of `wrapper`
        PyFrameObject *frame = PyEval_GetFrame();

        if (!tp_clear) {
            tp_clear = (frame->ob_type)->tp_clear;
        }
        // Only clear frames of initially obuscated functions
          while (tb) {
            PyObject *tb_frame = tb -> tb_frame;
            PyObject *f_code = *(PyObject**)((char*)tb_frame + f_code_offset);
            PyObject *co_code = *(PyObject**)((char*)f_code + co_code_offset);
            int co_flags = *(int*)((char*)f_code + co_flags_offset);

            if (co_flags & CO_PYARMOR_FLAG)
              tp_clear(tb_frame);
            tb = tb->tb_next;
          }
    }
}
```
The advantage is, that even callbacks from encrypted code to outside code are secured of getting its `f_locals`.

**Comment by jondy** [comment]
I think it's efficient and safe enough only clear these 2 frames. For example,
```python
def foo(filename):
    # It will raise OSError if pass a directory
    os.remove(filename)

@wraparmor
def level3(filename):
    foo(filename)

@wraparmor
def level2(filename):
    level3(filename)

@wraparmor
def level1(filename):
    level2(filename)
```
In outer script, call `level1`
```python
    level1('/')
```
It will raise OSError in frame `os.remove`. The frames of `level1`, `level2`, `level3` are `tp_clear`ed, `foo` and all the inner frame are not. Even if adding an extra flag to `foo`, all the inner frames of system library are still original. Besides, if iterating all the inner frames, it means many duplicated `tp_clear` in the same frame when `__wraparmor__(func, tb, 1)` is called:

* level3: clear nothing
* level2: clear `level3`
* level1: clear `level2`, `level3` (duplicated call)

So the ruler is to add decorator to any important function in python scripts.

**Comment by jondy** [comment]
Regard to initialize customized getter of `frame.f_locals` only when exception raised, what I thought is

* When debugging python scripts, every thing will be original if calling `__wraparmor__(func, None, 1)` at the end of `wrapper`

* In production mode, no exception raised, no frame can be accessed. So it need not to initialize this getter at the beginning.

**Comment by joonis** [comment]
I agree with clearing just the first two frames. It's sufficient.
But I don't like the approach that the getter is set only on exceptions. Have a look at the following example:

```py
def foo(filename):
    # It will raise OSError if pass a directory
    os.remove(filename)

@wraparmor
def level3(filename):
    foo(filename)

@wraparmor
def level2(filename):
    level3(filename)

@wraparmor
def level1(filename, callback):
    callback()
    level2(filename)
```

```py
def mycallback():
    frame = sys._getframe(1)
    print(frame.f_code.co_name, frame.f_locals)

level1('/', mycallback)
```
You would not get the locals of `level1` if the getter was initially set.
Maybe we should add a parameter to the first `__wraparmor__` call?

**Comment by jondy** [comment]
No any extra parameter required, just check `original_frame_getlocals` is `NULL` or not in the `__wraparmor__`:
```c
    if (!original_frame_getlocals)
      init_flocals_getter(frame);
```

**Comment by joonis** [comment]
Already implemented in v3.8.7?
It doesn't work. The example above still gives me the locals of `level1`.

**Comment by jondy** [comment]
Not yet, I find this issue just now. Fixed in v3.8.8

**Comment by joonis** [comment]
Unfortunately it doesn't work in v3.8.8 either.
The getter must be set in the first call of `__wraparmor__` which decrypts the bytecode.

**Comment by joonis** [comment]
Further more it should be possible to call `__wraparmor__(func, tb)` with just two arguments.
In this case the frames should be cleared but the code object should not be encrypted again.
I don't know how this is handled at time. Currently it already seems to work, but I don't know if frames are cleared correctly.

This is required to support dynamically created functions and it solves the problem with generator functions.

Are there any issues calling `__wraparmor__(func)` repeatedly without finalizing it
with `__wraparmor__(func, tb, 1)`?

**Comment by joonis** [comment]
Ok, encryption depends on `func.__refcalls__`. If I set it initially to `1` the bytecode is never encrypted again, right?

**Comment by jondy** [comment]
`func.__refcalls__` must be 0 initially. And `__wraparmor__(func, tb, 1)` must match `__wraparmor__(func)`.
Otherwise, pyarmor will think the function is running, so does nothing.
Call `__wraparmor__(func)` will increase `func.__refcalls`, and `__wraparmor__(func, tb, 1)` decrease.
Only `func.__refcalls__` equals 1, pyarmor goes into action.

Can still traceback get data from `frame.f_locals`?  I have a simple testcase, the traceback only get an empty dictionary.

**Comment by jondy** [comment]
Call `__wraparmor__(func, tb)` will not decrease `func.__refcalls`, so the function will not be encrypted again, and not `tp_clear` frame. But the customized getter takes effect, `frame.f_locals` still returns `{}`

**Comment by joonis** [comment]
If I call an external function from an encrypted function regardless of an exception,
the getter is not set and the external function can access the locals of the encrypted function.
Therefore the getter should be set initially or at least in the first call of `__wraparmor__(func)`.

Furthermore calling `__wraparmor__(func, tb, 1)` should clear the frames in any case, regardless of whether `func.__refcalls__` counts zero or not. Because the frames are independent from that value
and certainly finalized on exceptions. It's only important for the code object of the function.

**Comment by joonis** [comment]
The latter would also solve issues with generator functions, since the wrapper can initially set `func.__refcalls__` to `1`. So such functions won't be encrypted again but its locals are still protected.

**Comment by joonis** [comment]
Have a look at this example:
```py
        def wraparmor(func, shift_error=False, is_generator=False):
            func.__refcalls__ = is_generator and 1 or 0
            def wrapper(*args, **kwargs):
                __wraparmor__(func)
                tb = None
                try:
                    return func(*args, **kwargs)
                except Exception as err:
                    tb = exc_info()[2]
                    if shift_error:
                        raise err
                    raise
                finally:
                    __wraparmor__(func, tb, 1)
            return wrapper

        def wrapgenerator(func):
            return wraparmor(func, is_generator=True)

        @wrapgenerator
        def myfunc():
            for i in range(10):
                yield i
```
You cannot encrypt `myfunc` before the generator exits, otherwise it will cause a segfault.
`func.__refcalls__ = 1` solves this issue.

**Comment by joonis** [comment]
And here an example of the first issue.

The encrypted module:
```py
@wraparmor
def do_something(progress_callback):
    foo = 'secret'
    progress_callback()
```
End user script:
```py
from encrypted_module import do_something

def mycallback():
    frame = sys._getframe(1)
    print(frame.f_code.co_name, frame.f_locals)

do_something(mycallback)
```

**Comment by jondy** [comment]
No extra parameter `is_generator=False` required, it's better to tell generator in `__wraparmor__`:
```c
    if (co_flags & CO_GENERATOR) {
       // Do not obfuscate byte-code
    }
```
And `func.__refcalls__` always is set to 0 initially.

**Comment by joonis** [comment]
Even better, yes.

**Comment by jondy** [comment]
About the callback issue, I have almost same test case as your example, the callback can only get an empty dictionary in v3.8.8, does still the cached version takes effect?

**Comment by joonis** [comment]
No, version is 3.8.8.
I will try a simpler test case tomorrow...

**Comment by joonis** [comment]
Here it doesn't work. I still get the locals.
When is the getter set in v3.8.8?

**Comment by jondy** [comment]
The getter is set when any `__wraparmor__(func)` is called first time.

**Comment by jondy** [comment]
Here it's latest version 

http://pyarmor.dashingsoft.com/downloads/platforms/linux_x86_64/_pytransform.so
md5sum: 12d5e6b48ba58c11b00d8958f6580933

In this version
* Generator will not be obfuscated in any case
* Frame will be clear as long as `tb` is not None

I need to be sure the callback issuse is fixed before publish.

**Comment by joonis** [comment]
Latest version does work!
There is just one last issue with some nested decorators.
But I don't get it breaked down to a small piece of code...

**Comment by joonis** [comment]
Ok, I got it. The following example causes a segfault:

```py
def factory():
    
    @wraparmor
    def nestedfunc(f):
        f()
  
    return nestedfunc

func1 = factory()
func2 = factory()
func1(func2)
```
The problem is, that the codeobject of `nestedfunc` is always the same. So `func1` and `func2` hold a reference to the same codeobject. But we use a reference counter for each function instead the codeobject. Don't know if we could misuse some attribute of the codeobject for that ...

**Comment by jondy** [comment]
Fixed in v3.8.10. Use `co_names->ob_refcnt` as counter of each function call.

**Comment by joonis** [comment]
So `func.__refcalls__` is no longer required?

**Comment by jondy** [comment]
It's still required in v3.8.10. But it can be removed as long as `co_names->ob_refcnt` is stable in next version.

**Comment by joonis** [comment]
The example above works now. But there is still an issue with a similar construct. Strange.
For what is `func.__refcalls__` used at the moment?

**Comment by jondy** [comment]
I'm afraid `co_names->ob_refcnt` does not work in complex application. So keep `func.__refcalls__` temporary.

**Comment by joonis** [comment]
Since v3.9 it runs into a segfault again, even if obfuscated without `--obf-code-mode=wrap`.

**Comment by jondy** [comment]
Maybe it's better use the previous version of Pyarmor, only update the latest dynamic library `_pytransform.so`

**Comment by joonis** [comment]
Just updating the dynamic libraries results in a segfault as well.
Same with ` --obf-code-mode=des` and v3.9
